### PR TITLE
Binary metrics protocol

### PR DESCRIPTION
**Pull request by smoldachev**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -478,15 +478,15 @@
                                             <include name="**/*.sh" />
                                         </fixcrlf>
 
-                                        <zip destfile="${basedir}/target/bin/${ignite.edition}-${project.version}-bin.zip" encoding="UTF-8">
-                                            <zipfileset dir="${basedir}/target/release-package-${ignite.edition}" prefix="${ignite.edition}-${project.version}-bin" filemode="755">
+                                        <zip destfile="${basedir}/target/bin/${ignite.edition}-${project.version}.zip" encoding="UTF-8">
```
Comment (nva) [inline pom.xml]: Why we change this?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchema.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * Format:
+ * 0 - int - schema size in bytes.
+ *
+ * 4 - byte - value type
+ * 5 - int - name size (k)
+ * 9 - byte[k] - name bytes
+ * ... repeat ...
+ */
+//TODO: add version to the metric registry schema (should also contain node version for mixed clusters)
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchema.java]: TODO without issue number.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchemaItem.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        MetricRegistrySchemaItem that = (MetricRegistrySchemaItem) o;
+
+        return name.equals(that.name) && metricType == that.metricType;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int hashCode() {
+        return Objects.hash(name, metricType);
+    }
+}
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchemaItem.java]: Do we need `toString()` method for this class?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRequest.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+package org.apache.ignite.internal.processors.metric.export;
+
+import org.apache.ignite.plugin.extensions.communication.Message;
+import org.apache.ignite.plugin.extensions.communication.MessageReader;
+import org.apache.ignite.plugin.extensions.communication.MessageWriter;
+
+import java.nio.ByteBuffer;
+
+public class MetricRequest implements Message {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRequest.java]: Missing javadoc for class.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRequest.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    /** {@inheritDoc} */
+    @Override public byte fieldsCount() {
+        return 3;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onAckReceived() {
+        // No-op.
+    }
+}
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRequest.java]: Do we need `toString()` method for this class?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+ * Data:
+ * Accordingly to schema.
+ *
+ */
+public class MetricResponse implements Message {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** Protocol version 1. */
+    private static final short PROTO_VER_1 = 1;
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java]: why we have private static const with protocol version in several places?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+     * @return Schema version.
+     */
+    public int schemaVersion() {
+        return getInt(body, BYTE_ARR_OFF + SCHEMA_VER_OFF);
+    }
+
+    /**
+     * @return Cluster ID.
+     */
+    //TODO: could be null?
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java]: TODO without issue number.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+    /** {@inheritDoc} */
+    @Override public byte fieldsCount() {
+        return 1;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onAckReceived() {
+        // No-op.
+    }
+}
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java]: Do we need `toString()` method for this class?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricSchemaItem.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        MetricSchemaItem that = (MetricSchemaItem) o;
+
+        return idx == that.idx && pref.equals(that.pref);
+    }
+
+    /** {@inheritDoc} */
+    @Override public int hashCode() {
+        return Objects.hash(idx, pref);
+    }
+}
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricSchemaItem.java]: Do we need `toString()` method for this class?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+            ensureCapacity(2);
+
+            arr[pos++] = (byte)((val & 0x7F) | 0x80);
+
+            arr[pos++] = (byte)(val >>> 7);
+        }
+        else if ((val & (~0 << 21)) == 0) {
+            ensureCapacity(3);
+
+            arr[pos++] = (byte)((val & 0x7F) | 0x80);
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java]: Does it make sense to declare named constants for: 0x7F and 0x80 ?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+        arr[pos++] = (byte)(v >> 32);
+        arr[pos++] = (byte)(v >> 24);
+        arr[pos++] = (byte)(v >> 16);
+        arr[pos++] = (byte)(v >> 8);
+        arr[pos++] = (byte)(v);
+    }
+
+    /**
+     * Puts integer value in varint format. It can consume from 1 to 5 bytes.
+     *
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java]: this method is a very tricky, I think it make sense to provide more info about its implementation in javadocs.
May be a link to some internet resource about var int?
Like: https://learnmeabitcoin.com/guide/varint

File: parent/pom.xml [hunk 12]
```diff
@@ -552,6 +552,11 @@
                                 <packages>org.apache.ignite.spi.tracing.opencensus*</packages>
                             </group>
                         </groups>
+                        <header>
+                            <![CDATA[
```
Comment (akuznetsov-gridgain) [inline parent/pom.xml]: Why pom.xml was changed? It is not related to this issue.

File: pom.xml (hunk 1 above)
Comment (agoncharuk) [reply pom.xml]: By mistake a commit after local rebuild. Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchema.java (hunk 2 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchema.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchemaItem.java (hunk 3 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRegistrySchemaItem.java]: Added toString()

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRequest.java (hunk 4 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricRequest.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java (hunk 6 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java]: Removed this one, added a reference to request.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java (hunk 7 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java]: Removed the TODO

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java (hunk 8 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricResponse.java]: Added toString()

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricSchemaItem.java (hunk 9 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/MetricSchemaItem.java]: Added toString()

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java (hunk 10 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java]: Not quite sure - these are common masks for bit twiddling

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java (hunk 11 above)
Comment (agoncharuk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/export/VarIntWriter.java]: Added a short description to the class header.

File: parent/pom.xml (hunk 12 above)
Comment (agoncharuk) [reply parent/pom.xml]: A mistake - removed this one.

**Pull request closed without merging**